File: album-reviews.js
// All the revisions to this object
// will need to go through user-library check first
// and save to the whole database here

const albumReviews = {}; // contains all reviews for all albums
const allReviews = {}; // contains all reviews for a single album

albumReviews.getAllReviews = function getAllReviews(albumId) {
    return allReviews[albumId];
};

albumReviews.addReview = function addReview(albumId, review) {
    // don't need to return anything 
    // won't affect the current user's other reviews
    if (!allReviews[albumId]) {
        allReviews[albumId] = {}; // open a new slot for this album
    };

    allReviews[albumId][review.id] = review;
};

albumReviews.updateReview = function updateReview(albumId, review) {
    // don't need to return anything 
    // won't affect the current user's other reviews
    allReviews[albumId][review.id] = review;
};

albumReviews.deleteReview = function deleteReview(albumId, review) {
    delete allReviews[albumId][review.id];
};

module.exports = albumReviews;

File: backend/api/v2/pkg/models/album_reviews.go
package models

import (
	"sync"
)

type AlbumReviews struct {
	allReviews map[string]map[string]Review
	mu         sync.RWMutex
}

func NewAlbumReviews() *AlbumReviews {
	return &AlbumReviews{
		allReviews: make(map[string]map[string]Review),
	}
}

func (ar *AlbumReviews) GetAllReviews(albumId string) map[string]Review {
	ar.mu.RLock()
	defer ar.mu.RUnlock()
	return ar.allReviews[albumId]
}

func (ar *AlbumReviews) AddReview(albumId string, review Review) {
	ar.mu.Lock()
	defer ar.mu.Unlock()
	if _, exists := ar.allReviews[albumId]; !exists {
		ar.allReviews[albumId] = make(map[string]Review)
	}
	ar.allReviews[albumId][review.ID] = review
}

func (ar *AlbumReviews) UpdateReview(albumId string, review Review) {
	ar.mu.Lock()
	defer ar.mu.Unlock()
	if _, exists := ar.allReviews[albumId]; exists {
		ar.allReviews[albumId][review.ID] = review
	}
}

func (ar *AlbumReviews) DeleteReview(albumId string, reviewID string) {
	ar.mu.Lock()
	defer ar.mu.Unlock()
	if _, exists := ar.allReviews[albumId]; exists {
		delete(ar.allReviews[albumId], reviewID)
	}
}
